capi-configuration/README.md

```diff
@@ -0,0 +1,333 @@
+For example the `CIDR` in which the cluster is created.
+We currently assume that these values will be defaulted by custom webhooks or operators with significant provider specific domain knowledge.
+
+### Putting together a Giant Swarm release
+
+A Giant Swarm releases is just a set of `cluster-api` base files as defined previously with added `apps`.
+
+The following folder structure would be contained in `giantswarm/example-customer`:
+```yaml
+aws:
+  v21.0.0:
+    base:
+      cluster.yaml # configmap containing the base file (defined above)
+      aws-cluster.yaml
+      machinepool.yaml
+      awsmachinepool.yaml
+      ...
+      net-exporter.yaml # configmap containing version information for defaulting!
+    overlay:
+      special-overlay.yaml # Added by customer/AE/SA to e.g. overlay a different flatcar version!
+  v22.0.0:
+    base:
+      ...
+    overlay:
+      ...
+```
+
+In the management cluster a release is then simply a set of configmaps.
+The configmaps for e.g. release `v21.0.0` will be applied through argoCD by taking the `base` configmaps, applying any `overlay` from a customer and then deploying them.
+
+As shown in previous examples these configmaps would contain the release in their name (e.g. `machinepool-21-0-0`).
+Any in-cluster webhook or operator can then easily navigate to these configmaps to look up defaults for a cluster.
+
+A cluster creation workflow can now look as follows:
+1. Client (e.g. `happa`) submits `--dry-run` with `fundamental` custom resources to management cluster.
+2. Mutating webhooks take information from `base file` configmaps to default.
+3. Client receives fully defaulted custom resources and allows users to make manual changes.
+4. Client submits custom resources
+5. Cluster is created.
+
+A cluster upgrade workflow can then look like this:
+1. Client changes cluster version label.
+2. `upgrade-operator` reads `base file` configmaps according to the new version.
+3. `upgrade-operator` starts replacing fields with new configuration from configmaps.
+
+## Going gitops all the way
+
+A next step in `gitops` adoption would be the addition of having the cluster custom resources in `gitops` as well.
+So the shared git repository would contain not only release information but also the clusters themselves.
+
+```yaml
+gauss: # installation
+  peu01: #clusterID
+    base: # managed fully by GS
+      cluster.yaml
+      ...
+      cert-exporter.yaml # default app
+    overlay: # overwrites by customer
+      cluster.yaml
+      ...
+      kong.yaml # non default app
+aws: # releases as mentioned before
+  v21.0.0:
+  ...
+```
+
+Interaction with this repository could then be improved by adding functionality to our CLI `kubectl-gs`.
+
+Cluster creation would become a `pull request` which commits code generated by `kubectl-gs`.
+
+Upgrading a cluster is a bit more complex: Upgrades become a overwrite of configuration from the release to the `base` folder.
```

Comment:
This part does not make sense to me. IMO a base for a given release should never change. Wouldn't an upgrade be a change to the cluster resources?

Keeping the approach from above, wouldn't an upgrade be something like:
0. Client fetches the resource to be upgraded (from cluster or from git)
1. Client submits `--dry-run` with the new value(s) to the API server
2. Mutating webhooks apply the new defaults (like upgrade-operator)
3. Client receives fully defaulted custom resources and allows users to make manual changes.
4. User opens PR with changes to cluster resources
5. PR is merged, cluster is created.

Reply:
There is two `base` folders:
The release has a `base` with optional `overlay`
The cluster also has a `base` with optional `overlay`

What I describe here is an update of the `base` for the cluster - essentially taking the result of `release base` + `release overlay` and applying it to `cluster base` while `cluster overlay` stays unchanged.

The webhook workflow does not work like that unfortunately. If you submit a CR with non empty fields it will become undecidable if the customer e.g. wants to have a special setting in their cluster. The webhook would need to either overwrite all values (losing any custom setting for the cluster) or touch no already set values (essentially not upgrading).

Maybe we can discuss how to clarify this better.

Reply:
I don't see how the `rel base` + `rel overlay` solves the undecidability issue though if using webhooks, since the CR will still hit the same webhooks once applied. I feel like in either case we'd have to perform the defaulting and apply the resources (bypassing the defaulting) in separate steps. 

Thinking about other UX, we might have to somehow track which fields were set by the user anyway. For example so that from a CLI tool I can just change the cluster version and all necessary additional defaulting happens behind the scenes.

Reply:
It still isn't clear to me how the webhooks will be able to differentiate between fields they should mutate or not during an upgrade. Maybe the idea is to rely on [server-side apply](https://opensource.googleblog.com/2021/10/server-side-apply-in-kubernetes.html)?

Reply:
We are exploring different ideas here for the upgrades. We focus on creating releases and versioned clusters for now but there will be a follow up spec to nail down `upgrades` once we have more experience.

The current slack channel is `wg-capi-releases` - we have not checked `server-side apply` as described in the article yet.